halo2-base/src/safe_types/mod.rs

```diff
@@ -0,0 +1,133 @@
+pub use crate::{
+    gates::{
+        flex_gate::GateInstructions,
+        range::{RangeChip, RangeInstructions},
+    },
+    utils::ScalarField,
+    AssignedValue, Context,
+    QuantumCell::{self, Constant, Existing, Witness},
+};
+use std::cmp::{max, min};
+
+type RawAssignedValues<F> = Vec<AssignedValue<F>>;
+
+const BITS_PER_BYTE: usize = 8;
+// Each AssignedValue can at most represent 8 bytes.
+const MAX_BYTE_PER_ELEMENT: usize = 8;
+
+// SafeType's goal is to avoid out-of-range undefined behavior.
+// When building circuits, it's common to use mulitple AssignedValue<F> to represent
+// a logical varaible. For example, we might want to represent a hash with 32 AssignedValue<F>
+// where each AssignedValue represents 1 byte. However, the range of AssignedValue<F> is much
+// larger than 1 byte(0~255). If a circuit takes 32 AssignedValue<F> as inputs and some of them
+// are actually greater than 255, there could be some undefined behaviors.
+// SafeType gurantees the value range of its owned AssignedValue<F>. So circuits don't need to
+// do any extra value checking if they take SafeType as inputs.
+#[derive(Clone, Debug)]
+pub struct SafeType<F: ScalarField, const B: usize, const L: usize> {
+    value: RawAssignedValues<F>,
```

Comment:
I might be incurring future wrath, but for optimization should we make `value: [AssignedValue<F>; L]`. This avoids overhead from vector and heap.